Add read_number public functions

pavelxdd opened the pull request:
I really like the yyjson `read_number` implementation and want to use it in my projects. The problem is that the current implementation is a static function which can't be used from other code.

I've added a simple wrapper `unsafe_yyjson_read_number` around `read_number` function, which parses number from the memory buffer into `yyjson_val` (or `yyjson_mut_val`) variable.

ibireme commented on src/yyjson.h:
```diff
@@ -799,6 +801,48 @@ yyjson_api_inline size_t yyjson_read_max_memory_usage(size_t len,
+
+ This function is thread-safe when data is not modified by other threads.
+
+ @param dat The JSON data (UTF-8 without BOM), null-terminator is not required.
+    If this parameter is NULL, the function will fail and return false.
+ @param len The length of JSON data in bytes.
+    If this parameter is 0, the function will fail and return false.
```
The null-terminator is required for the internal implementation of `read_number`.
The parameter `len` is not required.

pavelxdd replied:
Yes, I've updated the comment message.
Not only the null-terminator is required, for example `123qwe` corretly parsed into `123` and `end` pointer is set to `q` character.

ibireme commented on src/yyjson.c:
```diff
@@ -5794,6 +5794,15 @@ yyjson_doc *yyjson_read_file(const char *path,
+    if (unlikely(!dat || !len)) return false;
+    hdr = (u8 *)dat, end = hdr + len;
+    if (!read_number(&hdr, NULL, false, (yyjson_val *)val, &msg)) return false;
+    return hdr == end;
```
Here we can return the end of the string directly, which provides more information to the caller (as `strtod()` does).

pavelxdd replied:
I've updated the commit to return the end pointer instead of a boolean.

pavelxdd commented:
I've updated the commit with following changes:

1) Removed `unsafe_yyjson_read_number` in favor of `yyjson_read_number`.
`yyjson_mut_read_number` is just a wrapper arount `yyjson_read_number`.

2) Return `end` pointer instead of boolean if parsing was successful, or return NULL instead.

3) Support `yyjson_read_err` optional parameter.

pavelxdd commented:
I've added test cases based on `test_number` text files.

ibireme commented:
Thanks, it looks much better now~
